lib/grandpa/votes_tracker_test.go

```diff
@@ -70,7 +70,9 @@ func Test_newVotesTracker(t *testing.T) {
 	}
 	vt := newVotesTracker(capacity)
 
-	assert.Equal(t, expected, vt)
+	assert.Equal(t, expected.mapping, vt.mapping)
+	assert.Equal(t, expected.linkedList, vt.linkedList)
+	assert.Equal(t, expected.capacity, vt.capacity)
 }
 
```

Comment:
It would be nice to add a test ensuring thread safety. I wrote one up for the votes tracker, it can be adapted as well for the commits tracker:

```go
func Test_votesTracker_threadSafety(t *testing.T) {
	// This test is meant to be run with the `-race` flag
	// to detect any data race.
	t.Parallel()

	const capacity = 2
	votesTracker := newVotesTracker(capacity)

	const parallelism = 4

	var startWg, endWg sync.WaitGroup
	startWg.Add(parallelism)
	endWg.Add(parallelism)

	timedOut := make(chan struct{})
	go func() {
		startWg.Wait()
		const duration = 50 * time.Millisecond
		time.Sleep(duration)
		close(timedOut)
	}()

	for i := 0; i < parallelism; i++ {
		go func(i int) {
			defer endWg.Done()
			startWg.Done()

			const peerID = peer.ID("abc")
			blockHash := common.Hash{byte(i)}
			voteMessage := &VoteMessage{
				Message: SignedMessage{
					BlockHash: blockHash,
				},
			}

			for {
				select {
				case <-timedOut:
					return
				default:
				}

				votesTracker.add(peerID, voteMessage)
				votesTracker.delete(blockHash)
				_ = votesTracker.messages(blockHash)
				_ = votesTracker.networkVoteMessages()
			}
		}(i)
	}

	endWg.Wait()
}

```

Reply:
@qdm12 Thanks for this. Adding this. 
It doesn't look like we run tests with `-race` automatically in CI. I only see `test-state-race`, which tests only for `dot/state`. I found similar data race tests in `dot/sync` and `lib/blocktree` as well. It doesn't look like they are running with `-race` at the moment. 

If that makes sense to you, I can create a separate small PR, adding  a make target say `test-race` and run those in CI as well.

Reply:
Yes that's true, the problem is `-race` makes tests a lot slower and use more memory, so we should really reserve it for unit tests. Now a bunch of unit tests are really integration tests, but it should be fixed more or less in #2887. What you can do is create an issue (or draft PR up to you) to add some `go test -race` workflows, once #2887 is merged?

Reply:
Created an issue https://github.com/ChainSafe/gossamer/issues/2946